### ISSUE TITLE
Allow completers to select text

### ISSUE THREAD

**Issue by refi64** [howl-editor/howl#144]
In `ui/completion_popup.moon`, there's this line:

``` moonscript
@editor.cursor.pos = @completer\accept item, @editor.cursor.pos
```

However, this makes it impossible for a completer to do stuff like select the inserted text.

My idea is to change that to:

``` moonscript
old_persistent = @editor.cursor.selection.persistent
@editor.cursor.selection.persistent = true
@editor.cursor.pos = @completer\accept item, @editor.cursor.pos
@editor.cursor.selection.persistent = old_persistent
```

(Of course, with proper error handling and stuff.) Thoughts?

**Comment by nilnor** [comment]
The current structure of this is that completers only complete, they are not expected to do anything else. To better understand why you want this - what is the use case you're thinking of here?

**Comment by kirbyfan64** [comment]
@nilnor In [howl-autoclang](https://github.com/kirbyfan64/howl-autoclang), placeholder text is inserted in `finish_completion`. For instance, if someone selected `puts` from the completion list, the text `(const char* something)` or similar will be inserted afterwards. Although this is useful, it's also slightly irritating to always have to delete the text inside the parens. So I figured I could select the text inside automatically, that way a simple backspace will get rid of it, but it's still there.

**Comment by shalabhc** [comment]
This kind of feature sounds awesome and I'm excited to see C/C++ autocompletion underway :+1: 

But is selection the right thing for this? Presumably one could insert some text with multiple placeholders that the user replaces one-by-one, by tabbing through them, for example. Also it might be useful to let the user replace the placeholder at a later time, something a selection won't do.

**Comment by nilnor** [comment]
Yes, I see the use for this. As Shalabh is suggesting, I don't think the selection is necessarily the best option for handling this. And even if is, I rather introduce an explicit way to assign this kind of information with a completion item itself, rather than having a completer do any modification itself. 

I was also going to bring up the introduction of placeholders - with the advent of aullar and buffer markers I think we now have a pretty good base to build on for creating a more general snippet expansion functionality. And this seems like it could be one such case.

**Comment by shalabhc** [comment]
@kirbyfan64 - see if the table-completions branch helps you a litte bit here. Specifically instead of returning completions as `{'text-to-insert1', 'text-to-insert2'}`, you can now return a list of tables like `{ {'text-to-display1',  completion: 'text-to-insert1'}, ...}` etc. You can even have multiple columns in the table if that helps with formatting.

**Comment by nilnor** [comment]
Sorry for the very long silence here.

Howl development has been quiet on `master` for years, while work continued on the `next` branch — that branch has now been merged. Master has changed substantially since the 0.6 release: a full port to GTK 4, and a rewrite of the command line and interaction APIs, among a lot else.

We're closing out the old backlog, and feature requests are being closed for good rather than parked. Howl is developed at a pace that doesn't support an open wishlist, and leaving these sitting here indefinitely implies a roadmap that doesn't exist. That's not a comment on the merit of the idea — thank you for taking the time to write it up.

Please don't reopen this one. Bug reports against current master are still welcome.
